src/bluesky/callbacks/tiled_writer.py

```diff
@@ -75,12 +78,16 @@ def _ensure_resource_backcompat(self, doc: StreamResource) -> StreamResource:
 
         doc = copy.copy(doc)
         if ("mimetype" not in doc.keys()) and ("spec" not in doc.keys()):
-            raise RuntimeError("StreamResource document is missing a mimetype or spec")
+            raise RuntimeError("StreamResource document is missing a 'mimetype' or 'spec'")
         else:
-            doc["mimetype"] = doc.get("mimetype") or MIMETYPE_LOOKUP[doc.get("spec")]
+            doc["mimetype"] = doc.get("mimetype") or MIMETYPE_LOOKUP[str(doc.get("spec"))]
         if "parameters" not in doc.keys():
             doc["parameters"] = doc.pop("resource_kwargs", {})
         if "uri" not in doc.keys():
+            if "root" not in doc.keys():
+                raise RuntimeError("StreamResource document is missing a 'root' path")
+            if "resource_path" not in doc.keys():
+                raise RuntimeError("StreamResource document is missing a 'resource_path'")
```

Comment:
Since this method functions as a converter from `Resource | StreamResource -> StreamResource` (not actually acting on a `StreamResource`) we could probably clean things up by putting these checks at the top of the method. It either contains a `"mimetype"` in which case it's a stream resource and we return the copy as is, or it has to contain `"spec"`, `"root"`, `"uri"`, `"resource_path"`, etc (anything we need for the conversion).

We should also change the name to `_convert_resource_to_stream_resource`.

Reply:
Ah nevermind, it does act on `StreamResource` it has to convert the [older version stream resource](https://github.com/bluesky/event-model/blob/v1.20.0/event_model/documents/stream_resource.py) to the [newer one](https://github.com/bluesky/event-model/blob/v1.22.3/src/event_model/documents/stream_resource.py).

The above suggestion still works though, if it has a `"mimetype"` it is a newer `StreamResource`, if it has a `"spec"` it is an older stream resource, or resource, and needs converting.

Reply:
Moving the runtime checks to the top :+1: 

I realize the `"mimetype"` check should be sufficient to assume that `doc` is a _new_ `StreamResource` but do we want to be explicit that it should have all 3 of `"mimetype"`, `"parameters"`, and `"uri"`?

Reply:
Just checking `"spec"` should be fine, I would be tempted to add an

``` python
assert {"mimetype", "parameters", "uri"} <= doc.keys()
```

but I don't think we need to validate that it is a stream resource > v1.20.0 here. We can assume at this point that we have a stream resource or resource adhering to the schema (which will have different definitions for stream resource depending on event-model version). Adding a check like the above would just leave more to change if the definitions of stream resource are altered.

I think all this needs is the current pops for outdated fields if present. Then putting them all in the same `StreamResource` cache outside the method.